### PR TITLE
Some Bug fix in JSON type casting  (Date, Interval, Timestamp), Improvement (literal NULL/NAN/INF) in CSV and new tests on datatype casting for JSON

### PR DESCRIPTION
**Pull request by SterlingT3485**
# Description

1. Fix the casting problem in JSON extension for `DATE, INTERVAL, TIMESTAMP, UUID` types 
     by rechecking the content of `YYJSON_TYPE_STR` in `jsonSchema` function of `json_utils.cpp`

2. Imporve the ability of handling the literal `NULL` in CSV, the previous version cannot handle the nested List/Array/Map with NULL, e.g.
`"[[a, NULL, c], NULL, [a, NULL, c]]"`
it would detect this type as `String[]`, instead of `String[][]`.
By adding a `NULL_STR` logic type for the literal `NULL` in CSV, now we can handle this.

3. Add multiple tests focus on the int and max value of each single datatype, as well as some complex hybrid nested types.
 

Fixes 4462

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: src/processor/operator/persistent/reader/csv/base_csv_reader.cpp [hunk 1]
```diff
@@ -279,6 +279,11 @@ uint64_t BaseCSVReader::parseCSV(Driver& driver) {
     value_start:
         // state: value_start
         // this state parses the first character of a value
+        // skip the spaces after delimiter
+        if (buffer[position] == ' ') {
```
Comment (acquamarin) [inline src/processor/operator/persistent/reader/csv/base_csv_reader.cpp:283]: I would suggest using `StringUtils::isSpace` instead.

File: src/function/cast_string_non_nested_functions.cpp [hunk 2]
```diff
@@ -172,6 +172,9 @@ LogicalType inferMinimalTypeFromString(std::string_view str) {
     if (cpy.size() == 0) {
         return LogicalType::ANY();
     }
+    if (cpy == "NULL") {
```
Comment (acquamarin) [inline src/function/cast_string_non_nested_functions.cpp:175]: I prefer not adding a new datatype, instead we treat it as `LogicalType::ANY()`

File: src/processor/operator/persistent/reader/csv/base_csv_reader.cpp (hunk 1 above)
Comment (SterlingT3485) [reply src/processor/operator/persistent/reader/csv/base_csv_reader.cpp:283]: discussed with chang, this feature is against CSV standard, removed.

File: src/function/cast_string_non_nested_functions.cpp (hunk 2 above)
Comment (SterlingT3485) [reply src/function/cast_string_non_nested_functions.cpp]: done this

File: dataset/all_types/duckdb_nested_types.csv [hunk 3]
```diff
@@ -0,0 +1,3 @@
...
+"[42, 999, NULL, NULL, -42]","[42.0, nan, inf, -inf, NULL, -42.0]","[1970-01-01, infinity, -infinity, NULL, 2022-05-12]","[1970-01-01 00:00:00, infinity, -infinity, NULL, 2022-05-12 16:23:45]","[1969-12-31 19:00:00-05, infinity, -infinity, NULL, 2022-05-12 19:23:45-04]","[🦆🦆🦆🦆🦆🦆, goose, NULL, ]","[[], [42, 999, NULL, NULL, -42], NULL, [], [42, 999, NULL, NULL, -42]]","{'a': 42, 'b': 🦆🦆🦆🦆🦆🦆}","{'a': [42, 999, NULL, NULL, -42], 'b': [🦆🦆🦆🦆🦆🦆, goose, NULL, ]}","[{'a': NULL, 'b': NULL}, {'a': 42, 'b': 🦆🦆🦆🦆🦆🦆}, NULL]","{key1=🦆🦆🦆🦆🦆🦆, key2=goose}",5,"[4, 5, 6]","[d, e, f]","[[4, 5, 6], [NULL, 2, 3], [4, 5, 6]]","[[d, e, f], [a, NULL, c], [d, e, f]]","[{'a': 42, 'b': 🦆🦆🦆🦆🦆🦆}, {'a': NULL, 'b': NULL}, {'a': 42, 'b': 🦆🦆🦆🦆🦆🦆}]","{'a': [4, 5, 6], 'b': [d, e, f]}","[[42, 999, NULL, NULL, -42], [], [42, 999, NULL, NULL, -42]]","[[4, 5, 6], [NULL, 2, 3], [4, 5, 6]]"
```
Comment (acquamarin) [inline dataset/all_types/duckdb_nested_types.csv:3]: Always add a newline at the end of files.

File: extension/json/test/copy_to_json.test [hunk 4]
```diff
@@ -4,11 +4,11 @@
 
 -CASE TinySnbCopyToJSON
 -SKIP_IN_MEM
--SKIP
+#-SKIP
```
Comment (acquamarin) [inline extension/json/test/copy_to_json.test:7]: Just delete this line.

File: extension/test_list [hunk 5]
```diff
@@ -0,0 +1,66 @@
+duckdb~test~test_files~remote_duckdb.AttachNotExistS3RemoteDB /home/sterling/Desktop/kuzu/extension/duckdb/test/test_files/remote_duckdb.test
```
Comment (acquamarin) [inline extension/test_list:1]: remove this file

File: src/function/cast_string_non_nested_functions.cpp [hunk 6]
```diff
@@ -165,11 +165,29 @@ static RE2& realPattern() {
     return retval;
 }
 
+bool isAny(std::string_view cpy) {
+    std::string cpy_upper = std::string(cpy);
```
Comment (acquamarin) [inline src/function/cast_string_non_nested_functions.cpp:169]: I would suggest renaming to `isAnyType`?

**Review by acquamarin — APPROVED** [review]

File: dataset/all_types/duckdb_nested_types.csv (hunk 3 above)
Comment (SterlingT3485) [reply dataset/all_types/duckdb_nested_types.csv:3]: added

File: extension/json/test/copy_to_json.test (hunk 4 above)
Comment (SterlingT3485) [reply extension/json/test/copy_to_json.test:7]: deleted

File: extension/test_list (hunk 5 above)
Comment (SterlingT3485) [reply extension/test_list:1]: removed

File: src/function/cast_string_non_nested_functions.cpp (hunk 6 above)
Comment (SterlingT3485) [reply src/function/cast_string_non_nested_functions.cpp:169]: renamed

**Pull request merged**